openwisp_users/tests/test_admin.py

```diff
@@ -169,6 +170,20 @@ def test_admin_change_user_email_empty(self):
         self.assertEqual(len(mail.outbox), 0)
         self.assertContains(response, 'errors field-email')
 
+    def test_admin_change_user_page_get_invalid_UUID(self):
+        admin = self._create_admin()
+        self.client.force_login(admin)
+        with self.subTest('Test for wrong identifier'):
+            response = self.client.get(
+                reverse(f'admin:{self.app_label}_user_change', args=['WRONG'])
+            )
+            self.assertEquals(response.status_code, 404)
+        with self.subTest('Test for non-existing user'):
+            response2 = self.client.get(
+                reverse(f'admin:{self.app_label}_user_change', args=[uuid.uuid4()],)
+            )
+            self.assertEquals(response2.status_code, 404)
+
```

Comment:
These tests will be needed to be updated too. Since. the user will be redirected to the index admin page, you will need to pass `follow=True` in `self.client.get` method so it follows the redirect. 

The assertions will now check whether similar text exists in the response:
`User with ID “wrong” doesn’t exist. Perhaps it was deleted?`

Reply:
There is an issue that I've been facing with the assertions in this case, it is because of the differences in the HTML response returned by Django 2.2 and Django 3.1. Django 2.2 returns the HTML in form of HTML entities, so a quote ` " ` is returned as `&quot;`, but it is in the form of a normal `"` in Django 3.1.

I tried to find a way around by decoding the HTML returned in Django 2.2 as:
```py
response.content = html.unescape(response.content.decode())
``` 
But it still doesn't seem to work.
Could you help me with it?

The subtest:
```py
response = self.client.get(
    reverse(f'admin:{self.app_label}_user_change', args=['WRONG']),
    follow=True,
)
content = 'User with ID'
response.content = html.unescape(response.content.decode())
self.assertIn(content, response.content.decode())
```

Reply:
@Shagunjain10 use [`assertContains`](https://docs.djangoproject.com/en/3.1/topics/testing/tools/#django.test.SimpleTestCase.assertContains) instead of `assertIn`. Read [Django testing documentation](https://docs.djangoproject.com/en/3.1/topics/testing/) to learn more writing tests.

Reply:
You probably meant to tag me XD.
I've used `assertContains` too, but the actual problem is `“` and `"`. You see there is this difference between the quotes, that is why the assertion fails. So, it's majorly because, different Django versions return different HTML entities in the response. :/

Reply:
@Saurav-Shrivastav You can test for different django versions accordingly:

```python
if django.VERSION < (3, 1):
  # check for version < 3.1
else:
  # check for other versions
```

Reply:
I am thinking of using regex for the same, though I don't have much idea about it, will probably read about it. Do you think that it is a viable approach?

Reply:
@Purhan , Won't that cause problems, when we shift to other versions of Django in future?

Reply:
> @Purhan , Won't that cause problems, when we shift to other versions of Django in future?

No, we can remove that when we drop support for some version, just make sure to leave a comment like this

```
#TODO: Remove this when we drop support for django<3.0
```

Reply:
Thanks for the help @pandafy and @Purhan. The PR now looks complete. Please let me know, if I've missed out on anything.